atlas/packages/components/src/components/Carousel/Carousel.tsx

```diff
@@ -1,39 +1,105 @@
-import React, { ReactNode, useRef } from "react";
+import React, { useState, useRef, useEffect, useCallback, useMemo } from "react";
+import { css, SerializedStyles } from "@emotion/core";
+import { animated, useSpring } from "react-spring";
+import useResizeObserver from "use-resize-observer";
 import { useCSS, CarouselStyleProps } from "./Carousel.style";
 import NavButton from "../NavButton";
 
 type CarouselProps = {
-	children: Array<ReactNode>;
-	scrollAmount?: Number;
+	children: React.ReactNode;
+	containerCss: SerializedStyles;
+	leftControlCss: SerializedStyles;
+	rightControlCss: SerializedStyles;
+	onScroll: (direction: "left" | "right") => void;
 } & CarouselStyleProps;
 
-export default function Carousel({ children, scrollAmount = 200, ...styleProps }: CarouselProps) {
-	const container = useRef(null);
-
-	function onScroll(direction: "right" | "left") {
-		container.current.scrollBy({
-			left: direction === "left" ? -scrollAmount : scrollAmount,
-			behavior: "smooth",
-		});
+const Carousel: React.FC<Partial<CarouselProps>> = ({
+	children,
+	containerCss,
+	leftControlCss,
+	rightControlCss,
+	onScroll = () => {},
+}) => {
+	if (!Array.isArray(children)) {
+		return <>{children}</>;
 	}
+	let [props, set] = useSpring(() => ({
+		transform: `translateX(0px)`,
+	}));
+	const [x, setX] = useState(0);
+	const { width: containerWidth, ref: containerRef } = useResizeObserver<HTMLDivElement>();
+	const elementsRefs = useRef<(HTMLDivElement | null)[]>([]);
+	const [childrenLength, setChildrenLength] = useState(0);
+	useEffect(() => {
+		elementsRefs.current = elementsRefs.current.slice(0, children.length);
+		const childrensLength = elementsRefs.current.reduce(
+			(accWidth, el) => (el != null ? accWidth + el.clientWidth : accWidth),
+			0
+		);
+		setChildrenLength(childrensLength);
+	}, [children.length]);
 
-	let styles = useCSS(styleProps);
-
+	const styles = useMemo(() => useCSS({}), []);
 	return (
-		<div css={styles.wrapper}>
-			<div ref={container} css={styles.container}>
-				{children.map((item, index) => (
-					<div key={`carousel-${index}`} css={styles.item}>
-						{item}
-					</div>
+		<div css={[styles.container, containerCss]}>
+			<div css={styles.itemsContainer} ref={containerRef}>
+				{children.map((element, idx) => (
+					<animated.div
+						style={props}
+						key={`Carousel-${idx}`}
+						ref={(el) => {
```

Comment:
I believe this can be just `ref={elementsRefs.current[idx]}` (not 100% sure)

Reply:
I don't think changing that would work, it gives the following TS error which is quite expressive:
```typescript
Type 'HTMLDivElement | null' is not assignable to type 'string | RefObject<HTMLDivElement> | ((instance: HTMLDivElement | null) => void) | null | undefined'.
  Type 'HTMLDivElement' is not assignable to type 'string | RefObject<HTMLDivElement> | ((instance: HTMLDivElement | null) => void) | null | undefined'.
    Type 'HTMLDivElement' is not assignable to type 'string'.ts(2322)
```

`elementsRefs.current` holds an array of `HTMLDivElements` and an `HTMLDivElement` can't be passed as a ref since react will only accept a ref object, a callback ref or null. 
I could make `elementsRefs` hold `RefObjects` instead of `HTMLDivElements`, but that has the price of making other parts of the code less readable. 
Not worth it imho.

Another possible way of shortening this is by changing it to:  
```javascript
<animated.div
     // ... Other props
     ref={el => (elementsRef.current[idx]=el)}
</animated.div>
```
Which is shorter, but harder to read / understand at a glance.

Reply:
You're right, we would need to hold `RefObject[]` instead, it may not be worth it.

Regarding the second part of your response, I think we should go that way. I actually believe `ref={el => (elementsRef.current[idx]=el)}` is easier to understand, since it does just one thing - assign the ref, we don't need to return it as well in that function, right?